File: go/tests/agriculture_board_test.go
// agriculture_board_test.go
//
// Verifies the CircleAI.Agriculture port (agriculture_board.go): field add/get,
// crops-for-field ordered by planting date, and average yield by variety
// (case-insensitive join, 0 when none).

package circleai_test

import (
	"math"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestAgriculture_FieldsAndCrops(t *testing.T) {
	b := circleai.NewInMemoryFarmBoard()
	b.AddField(circleai.Field{FieldId: "f1", AreaHa: 12.5, SoilType: "loam", IrrigationKind: "drip"})
	if got, ok := b.GetField("f1"); !ok || got.AreaHa != 12.5 {
		t.Fatalf("get field = %+v ok=%v", got, ok)
	}
	if _, ok := b.GetField("none"); ok {
		t.Fatalf("missing field found")
	}
	jan := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
	b.Plant(circleai.Crop{CropId: "c2", FieldId: "f1", Variety: "Maize", PlantedOn: jan.AddDate(0, 2, 0)})
	b.Plant(circleai.Crop{CropId: "c1", FieldId: "f1", Variety: "Maize", PlantedOn: jan})
	b.Plant(circleai.Crop{CropId: "c3", FieldId: "f2", Variety: "Wheat", PlantedOn: jan})

	crops := b.CropsForField("f1")
	if len(crops) != 2 || crops[0].CropId != "c1" || crops[1].CropId != "c2" {
		t.Fatalf("crops-for-field ordered by PlantedOn failed: %+v", crops)
	}
}

func TestAgriculture_AvgYield(t *testing.T) {
	b := circleai.NewInMemoryFarmBoard()
	jan := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
	b.Plant(circleai.Crop{CropId: "c1", FieldId: "f1", Variety: "Maize", PlantedOn: jan})
	b.Plant(circleai.Crop{CropId: "c2", FieldId: "f1", Variety: "maize", PlantedOn: jan})
	b.RecordYield(circleai.YieldRecord{CropId: "c1", TonsPerHa: 6, HarvestedOn: jan})
	b.RecordYield(circleai.YieldRecord{CropId: "c2", TonsPerHa: 8, HarvestedOn: jan})

	if avg := b.AvgYieldOfVariety("MAIZE"); math.Abs(avg-7.0) > 1e-9 {
		t.Fatalf("avg yield = %v, want 7", avg)
	}
	if avg := b.AvgYieldOfVariety("Sorghum"); avg != 0.0 {
		t.Fatalf("avg yield (none) = %v, want 0", avg)
	}
}

File: typescript/tests/agriculture_board.test.ts
// agriculture_board.test.ts
// Verifies the CircleAI.Agriculture port: fields, crops ordered by planting date,
// average-yield-of-variety rollup.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { InMemoryFarmBoard, AgricultureDomainContext, field, crop, yieldRecord } from "../src/agriculture/index";

describe("InMemoryFarmBoard", () => {
  it("adds and retrieves fields", () => {
    const b = new InMemoryFarmBoard();
    b.addField(field("f1", 12.5, "loam", "drip"));
    assert.equal(b.getField("f1")?.soilType, "loam");
    assert.equal(b.getField("nope"), undefined);
  });

  it("lists a field's crops earliest-planted first", () => {
    const b = new InMemoryFarmBoard();
    b.plant(crop("c3", "f1", "Maize", new Date("2026-03-01T00:00:00Z"), null));
    b.plant(crop("c1", "f1", "Maize", new Date("2026-01-01T00:00:00Z"), null));
    b.plant(crop("c2", "f1", "Wheat", new Date("2026-02-01T00:00:00Z"), null));
    b.plant(crop("cx", "f2", "Soy", new Date("2026-01-15T00:00:00Z"), null));
    assert.deepEqual(
      b.cropsForField("f1").map((c) => c.cropId),
      ["c1", "c2", "c3"],
    );
  });

  it("averages yield across a variety (case-insensitive), 0 when none", () => {
    const b = new InMemoryFarmBoard();
    b.plant(crop("c1", "f1", "Maize", new Date("2026-01-01T00:00:00Z"), null));
    b.plant(crop("c2", "f1", "maize", new Date("2026-01-01T00:00:00Z"), null));
    b.plant(crop("c3", "f1", "Wheat", new Date("2026-01-01T00:00:00Z"), null));
    b.recordYield(yieldRecord("c1", 6, new Date("2026-06-01T00:00:00Z")));
    b.recordYield(yieldRecord("c2", 8, new Date("2026-06-01T00:00:00Z")));
    b.recordYield(yieldRecord("c3", 4, new Date("2026-06-01T00:00:00Z")));
    assert.equal(b.avgYieldOfVariety("MAIZE"), 7);
    assert.equal(b.avgYieldOfVariety("Barley"), 0);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(AgricultureDomainContext.systemPromptSnippet.includes("[DOMAIN: Agriculture]"));
    assert.deepEqual(AgricultureDomainContext.complianceFlags, ["DAFF_regs", "CARA", "Fertilizer_Act", "POPIA"]);
    assert.deepEqual(AgricultureDomainContext.suggestedTools, [
      "weather_api",
      "market_prices",
      "soil_data",
      "document_editor",
    ]);
  });
});
